### PR TITLE
Fix TextSource incorrect handling in channels that return short reads.

### LINKED ISSUE #23375: [Bug]: TextSource fails to handle parsing channels that return very short reads (< # delimiter characters)
### What happened?

The `readDefaultLine` and `readCustomLine` methods do not handle cases where `0` or `1` length buffers are returned causing `IndexOutOfBoundsExceptions` due to incorrect calculation of the `appendLength`.

This was caused by https://github.com/apache/beam/commit/30a48f05cf2ee0eea0a304fea01eb40f323f9f3c and is unreleased at the moment.

### Issue Priority

Priority: 0

### Issue Component

Component: io-java-text

### PR DESCRIPTION
**Pull request by lukecwik**
The issue is that readDefaultLine and readCustomLine was incorrectly calculating the appendLength when the buffer returned was 0 length. This was solved by ensuring that the internal read loop always read at least one byte allowing for the code to ensure that we were making progress. For readDefaultLine we kept track of whether we need to skip an LF in the next buffer if the current buffer ended with a CR and for readCustomLine we had to remember how much of the delimiter we have read so far in this buffer.

The bug was introduced in https://github.com/apache/beam/commit/30a48f05cf2ee0eea0a304fea01eb40f323f9f3c

There was no noticeable change in the TextSourceBenchmark performance results.

Fixes #23375

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOReadTest.java [hunk 1]
```diff
@@ -363,10 +369,57 @@ public static Iterable<Object[]> data() {
...
+          getTextSource(path.toString(), null)
+              .createForSubrangeOfFile(metadata, 0, metadata.sizeBytes());
+      FileBasedReader<String> reader =
+          source.createSingleFileReader(PipelineOptionsFactory.create());
+      ReadableByteChannel channel =
+          FileSystems.open(
+              FileSystems.matchSingleFileSpec(source.getFileOrPatternSpec()).resourceId());
+      InputStream stream = Channels.newInputStream(channel);
+      reader.startReading(
+          new ReadableByteChannel() {
```
Comment (TheNeuralBit) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOReadTest.java:392]: nit: could be nice to document this for readability (could happen later though, I know you want to get this in)
```suggestion
      reader.startReading(
          // Placeholder channel that only yields 0- and 1-length buffers.
          // Data is read one byte at a time from line parameter
          new ReadableByteChannel() {
```

**Review by TheNeuralBit — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
It looks like the bug was caught with an internal Google test, please make sure that this patch fixes it before merging. Thanks!

**Comment by lukecwik** [comment]
> It looks like the bug was caught with an internal Google test, please make sure that this patch fixes it before merging. Thanks!

Will do.

**Comment by lukecwik** [comment]
Java_Examples_Dataflow_Java11 succeeded, github UI failed to update: https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Commit/12526/

**Comment by lukecwik** [comment]
Internal testing within Google also passed.

**Pull request merged**